src/lib/object.ts

```diff
@@ -4,7 +4,43 @@ import { UnexpectedError } from '../internal/errors';
  * Creates a deep clone of the value.
  */
 export function clone<T>(value: T): T {
-  return JSON.parse(JSON.stringify(value)) as T;
+  if (value === null) {
+    return value;
+  }
+
+  if (value instanceof Date) {
+    return new Date(value.getTime()) as unknown as T;
+  }
+
+  if (value instanceof Array) {
+    const arrayCopy = [] as unknown[];
+    value.forEach(item => {
+      arrayCopy.push(item);
+    });
+
+    return arrayCopy.map((item: unknown) =>
+      clone<unknown>(item)
+    ) as unknown as T;
```

Comment:
Can you avoid iterating the array twice? e.g.
```suggestion
   for (const item of value) {
     arrayCopy.push(clone<unknown>(item))
   }

   return arrayCopy as unknown as T;
```

(it's a pretty minimal optimization but for/of typically beats forEach performance without sacrificing readability)

Reply:
i think creating the copy array is unnecessary, something like `return value.map(clone)` could be enough, but typescript will probably shout about it